fpsd/features.py

```diff
@@ -0,0 +1,765 @@
+                 "  (SELECT *                                       "
+                 "  FROM crosstab(                                  "
+                 "    'SELECT                                       "
+                 "    exampleid, rank,                              "
+                 "     case when ingoing = true then 0 else 1 end   "
+                 "  FROM (                                          "
+                 "    SELECT                                        "
+                 "      ROW_NUMBER() OVER                           "
+                 "      (PARTITION BY exampleid ORDER BY t_trace)   "
+                 "      AS rank,                                    "
+                 "      t.*                                         "
+                 "    FROM raw.frontpage_traces t) x                "
+                 "  WHERE x.rank <= 10')                            "
+                 "  AS  ct(exampleid integer, {})                   "
+                 ");").format(', '.join(crosstab_columns))
+
+        self.execute_query_from_string(query)
+        return "features.initial_cell_directions"
+
+    def generate_table_outgoing_cell_ordering(self, num_features=500):
+        """This method takes all examples and produces a
+        table with the following format:
+
+        exampleid  | a         | b         ...    | num_features
+        (integer)  | (bigint)  | (bigint)  ...    | (bigint)
+        ---------------------------------------------------------
+        9          | 2         | 4         ...
+        10         | 1         | 2         ...
+        11         | 1         | 3         ...
+
+        where there are a variable number (num_features) of
+        feature columns, each named outgoing_cell_position_x
+        where x is the position in the trace, beginning at 1.
+
+        Args:
+            num_features [int]: number of cells to use as features
+
+        Returns:
+            [string] name of newly created table
+        """
+
+        self.drop_stale_feature_table("features.cell_ordering")
+
+        self._create_temp_packet_positions(outgoing_only=True)
+        self._create_table_outgoing_cell_positions(num_cells=num_features)
+
+        crosstab_columns = ['outgoing_cell_position_{} bigint'.format(x+1)
+                            for x in range(num_features)]
+
+        query = ("CREATE TABLE features.cell_ordering AS          "
+                 "(SELECT * FROM crosstab(                        "
+                 "'SELECT exampleid, outgoing_cell_position, rank "
+                 "FROM top_{}_outgoing_packet_positions')         "
+                 "AS ct(exampleid integer,                        "
+                 "{}));").format(num_features,
+                                 ', '.join(crosstab_columns))
+
+        self.execute_query_from_string(query)
+        return "features.cell_ordering"
+
+    def generate_table_outgoing_cell_ordering_differences(self,
```

Comment:
Yeah, I don't know where I was going with this one... Off-topic: I wish Github supported a basic, safe subset of LaTeX.